subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy

```diff
@@ -949,6 +949,77 @@ class ArtifactTransformIntegrationTest extends AbstractHttpDependencyResolutionT
         output.contains("files: [lib.jar]")
     }
 
+    def "transforms registering the input as an output can use normalization"() {
+        file("input1.jar").text = "jar"
+        file("input2.jar").text = "jar"
+        buildFile("""
+            import org.gradle.api.artifacts.transform.TransformParameters
+
+            configurations {
+                api1 {
+                    attributes { attribute usage, 'api' }
+                }
+                api2 {
+                    attributes { attribute usage, 'api' }
+                }
+            }
+
+            abstract class IdentityTransform implements TransformAction<TransformParameters.None> {
+                @Classpath
+                @InputArtifact
+                abstract Provider<FileSystemLocation> getInputArtifact()
+
+                void transform(TransformOutputs outputs) {
+                    println "Selecting input artifact \${inputArtifact.get().asFile}"
+                    outputs.file(inputArtifact)
+                }
+            }
+
+            dependencies {
+                registerTransform(IdentityTransform) {
+                    from.attribute(artifactType, 'jar')
+                    to.attribute(artifactType, 'transformed')
+                }
+            }
+
+            task producer1(type: Jar)
+            task producer2(type: Jar)
+            tasks.withType(Jar).configureEach {
+                destinationDirectory = layout.buildDirectory.dir("produced")
+                archiveBaseName = name
+            }
+
+            ["api1", "api2"].each { conf ->
+                tasks.register("resolve\$conf", Copy) {
+                    duplicatesStrategy = 'INCLUDE'
+                    def artifacts = configurations."\$conf".incoming.artifactView {
+                        attributes { it.attribute(artifactType, 'transformed') }
+                    }.artifacts
+                    from artifacts.artifactFiles
+                    into "\${buildDir}/libs1"
+                    doLast {
+                        println "files: " + artifacts.collect { it.file.name }
+                        println "ids: " + artifacts.collect { it.id }
+                        println "components: " + artifacts.collect { it.id.componentIdentifier }
+                        println "variants: " + artifacts.collect { it.variant.attributes }
+                    }
+                }
+            }
```

Comment:
I think we should invest some time into decreasing redundancy in this now-2703-line test class. There is just so much noise around the setup that it's really hard to see what exactly the test is trying to prove.

Reply:
I added a project builder test to only test the execution of the artifact tranforms. Please tell me what you think.